java/arcs/sdk/Particle.kt

```diff
@@ -13,5 +13,57 @@ package arcs.sdk
 
 /** Base interface for all particles. */
 interface Particle {
-    // TODO: Add methods common to wasm and jvm Particles here.
+    /**
+     * React to handle updates.
+     *
+     * Called for handles when change events are received from the backing store. Default action is to trigger
+     * rendering.
+     *
+     * @param handle Singleton or Collection handle
+     */
+    fun onHandleUpdate(handle: Handle) = Unit
+
+    /**
+     * React to handle synchronization.
+     *
+     * Called for handles that are marked for synchronization at connection, when they are updated with the full model
+     * of their data. This will occur once after setHandles() and any time thereafter if the handle is resynchronized.
+     * Default action is to trigger rendering.
+     *
+     * @param handle Singleton or Collection handle
+     * @param allSynced flag indicating if all handles are synchronized
+     */
+    fun onHandleSync(handle: Handle, allSynced: Boolean) = Unit
+
+    /**
+     * Rendering through UiBroker.
+     *
+     * Only implemented for wasm, no-op on JVM.
+     */
+    fun renderOutput() = Unit
+
+    /**
+     * Define template for rendering (optional).
+     *
+     * Only implemented for wasm, no-op on JVM.
+     *
+     * @param slotName name of slot where template is rendered.
+     * @see [renderOutput]
+     */
+    fun getTemplate(slotName: String): String? = null
```

Comment:
I had this debate myself. You're right, all of that stuff is DOM only, so won't be in the prod particles. However, it's very useful to compile these particles on the JVM, for a number of reasons. It allegedly helps IntelliJ understand the code (although this is broken now for me), but it will also let us more easily write unit tests. The other thing is that we have a large corpus of Kotlin particles that we can use as compile tests (all the tutorial ones, thanks @SHeimlich!) however, almost all of them use these DOM methods.

I decided it'd be better to provide these no-op methods on the base class, so that we can continue building all these particles on the jvm. Later on we can revisit.

Reply:
maybe we can add an DomParticle base class that implements Particle but also introduces the Dom specific methods. This is how the ts code is structured as well.

Reply:
Interesting idea.. I'll need to think about how that will interact with the generated particle base class, but it could be possible.